framework/validators/IpValidator.php

```diff
@@ -0,0 +1,555 @@
+     *
+     * @param $ip string
+     * @return string|array
+     *  string - the validation was successful;
+     *  array  - an error occurred during the validation.
+     * Array[0] contains the text of an error, array[1] contains values for the placeholders in the error message
+     */
+    public function validateSubnet($ip)
+    {
+        if (!is_string($ip)) {
+            return [$this->wrongIp, []];
+        }
+
+        $negation = null;
+        $cidr = null;
+        $isCidrDefault = false;
+
+        if (preg_match($this->getIpParsePattern(), $ip, $matches)) {
+            $negation = ($matches[1] !== '') ? $matches[1] : null;
+            $ip = $matches[2];
+            $cidr = isset($matches[4]) ? $matches[4] : null;
+        }
+
+        if ($this->subnet === true && $cidr === null) {
+            return [$this->noSubnet, []];
+        }
+        if ($this->subnet === false && $cidr !== null) {
+            return [$this->hasSubnet, []];
+        }
+        if ($this->negationChar === false && $negation !== null) {
+            return [$this->wrongIp, []];
+        }
+
+        if ($this->getIpVersion($ip) == 6) {
+            if ($cidr !== null) {
+                if ($cidr > static::IPV6_ADDRESS_LENGTH || $cidr < 0) {
+                    return [$this->wrongCidr, []];
+                }
+            } else {
+                $isCidrDefault = true;
+                $cidr = static::IPV6_ADDRESS_LENGTH;
+            }
+
+            if (!$this->ipv6) {
+                return [$this->ipv6NotAllowed, []];
+            }
+            if (!$this->validate6($ip)) {
+                return [$this->wrongIp, []];
+            }
+
+            if ($this->expandV6) {
+                $ip = $this->expand6($ip);
+            }
+        } else {
+            if ($cidr !== null) {
+                if ($cidr > static::IPV4_ADDRESS_LENGTH || $cidr < 0) {
+                    return [$this->wrongCidr, []];
+                }
+            } else {
+                $isCidrDefault = true;
+                $cidr = static::IPV4_ADDRESS_LENGTH;
+            }
+
+            if (!$this->ipv4) {
+                return [$this->ipv4NotAllowed, []];
+            }
+            if (!$this->validate4($ip)) {
+                return [$this->wrongIp, []];
+            }
+        }
+
+        if (!$this->isAllowed($ip, $cidr)) {
+            return [$this->notInRange, []];
+        }
+
+        $result = $negation . $ip;
+
+        if ($this->subnet !== false && (!$isCidrDefault || $isCidrDefault && $this->normalize)) {
+            $result .= "/$cidr";
+        }
+
+        return $result;
+    }
+
+    /**
+     * Expands an IPv6 address to it's full notation. For example `2001:db8::1` will be
+     * expanded to `2001:0db8:0000:0000:0000:0000:0000:0001`
+     *
+     * @param string $ip the original IPv6
+     * @return string the expanded IPv6
+     */
+    public static function expand6($ip)
+    {
+        $hex = unpack("H*hex", inet_pton($ip));
```

Comment:
It's better to use `'H*hex'` since you're not using any special chars or variables in the string.